refactors Journalist Interface for semantic HTML5/ARIA markup

cfm opened the pull request:
## Status

* Accessibility changes approved by @SaptakS 
* Ready for final maintainer's review
    * See "Checklist" section for pre-merge steps

## Description of Changes

Closes #5987 by:

1. [x] refactoring all `securedrop/journalist_templates/*.html` for semantic markup;
2. [x] adding ARIA annotations where HTML5 semantics are insufficient; and
3. [x] fixing SASS/CSS in light of (1)–(2), especially to move decorative images out of the DOM and accessibility trees.

## Testing

Click through the Journalist Interface and inspect in terms of:

1. ***Markup,* by inspecting and validating HTML.** This need not take place in Tor Browser; I used the [Web Developer extension](https://chrispederick.com/work/web-developer/) in Chrome.
2. ***Accessibility in assistive tech,* by listening to the interface as announced from Tor Browser by a screen-reader.**  It should sound and "feel" conspicuously more fluent.  (I also audited pages in [WebAIM's Wave Evaluation Tool](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh) and [Deque's axe DevTools](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd), but these will yield mixed results until the conclusion of this accessibility push.)
3. ***Visual appearance,* by comparing the interface against <https://demo-journalist.securedrop.org>.**  Changes are intended to be "pixel-perfect" except where there were preexisting idiosyncrasies that more-semantic markup tended to make naturally more consistent (or where noted in individual commits).  Of course, suggestions for fine-tuning are welcome!

Steps (2)–(3) should be performed in Tor Browser at both "standard" and "safest" security levels.

Finally:

4. Click through the Source Interface and compare against <https://demo-source.securedrop.org> for any regressions introduced by shared Sass.

## Deployment

No special considerations.

## Checklist

### After approval and before merge

1. [x] Reformat `securedrop/journalist_templates/*.html`
2. [x] Update as many translations as possible per <https://github.com/freedomofpress/securedrop/issues/5972#issuecomment-947317684>; test in Weblate sandbox  →  see <https://github.com/freedomofpress/securedrop/pull/6165#issuecomment-1016857219>
3. Approver will squash-merge ~(especially fix-ups) to one commit per pass~

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

cfm committed:
first pass through "/login"

cfm committed:
replicates changes from Source Interface's locale selector

Literally:

$ cp securedrop/source_templates/locales.html securedrop/journalist_templates/locales.html

cfm committed:
first pass through "/"

cfm committed:
first pass through "/": refactors source listing as TABLE

This change requires that utility functions such as show() use "display:
revert" rather than assuming "display: block" for non-hidden elements.

cfm committed:
first pass through "/col/<filesystem_id>": refactors p.breadcrumbs as UL

cfm committed:
first pass through "/col/<filesystem_id>": refactors collection listing as TABLE

cfm committed:
first pass through "/col/<filesystem_id>"

cfm committed:
first pass through "/account/account"

cfm committed:
first pass through "/admin"

cfm committed:
first pass through "/admin/add"

cfm committed:
first pass through "*modal.html": refactors as DIALOGs

cfm committed:
first pass through "/admin/config"

cfm committed:
first pass through HOTP/TOTP routes

cfm committed:
first pass through "/bulk"

cfm committed:
second pass through "/login": DIV#js-strings must be in BODY, not HEAD

cfm committed:
second pass through "/"

Of note, BUTTON is invalid inside of A.  However, the existing
fallback mechanism works as intended with just the A element present.

cfm committed:
second pass through "/col/<filesystem_id>"

cfm committed:
second pass through "/account/account"

cfm committed:
second pass through "/account/account": refactors CSS tooltips to look at custom "data-tooltip" attribute rather than nonexistent "tooltip"

cfm committed:
second pass through "/admin"

cfm committed:
second pass through "/admin/config"

cfm committed:
test /admin

cfm commented:
@SaptakS, while this pull request is officially still a draft, I wanted to request your accessibility review of the markup refactoring I've proposed here.  Per discussion with the SecureDrop team, once I've revised this draft based on your feedback, I'll make the CSS changes necessary for pixel-perfect consistency with `develop`, for final review by @zenmonkeykstop.

Note that one test (`functional/test_submit_and_retrieve_message.py::TestSubmitAndRetrieveMessage::test_submit_and_retrieve_happy_path`) is still flaking.  Once I've come up with a fix, I'll append that commit to this branch to avoid disrupting your review, Saptak, and then squash it in prior to merge.

cfm commented:
Confirmed that `functional/test_submit_and_retrieve_message.py::TestSubmitAndRetrieveMessage::test_submit_and_retrieve_happy_path` is indeed a flake, with local passes reproduced at last in CI as of <https://app.circleci.com/pipelines/github/freedomofpress/securedrop/3208/workflows/70299cd8-95de-48df-b79b-c7a4cf6bb764>.

cfm commented:
The flake, most demonstrably seen in <https://app.circleci.com/pipelines/github/freedomofpress/securedrop/3208/workflows/dd8b4348-b646-4cbb-8ac1-af76e1173290/jobs/58091/tests#failed-test-0>, looks like:

```
self = <tests.functional.test_submit_and_retrieve_message.TestSubmitAndRetrieveMessage object at 0x7ff49d42ebb0>

    def test_submit_and_retrieve_happy_path(self):
        self._source_visits_source_homepage()
        self._source_chooses_to_submit_documents()
        self._source_continues_to_submit_page()
        self._source_submits_a_message()
        self._source_logs_out()
        self.switch_to_firefox_driver()
        self._journalist_logs_in()
        self._journalist_checks_messages()
>       self._journalist_downloads_message()

/home/circleci/project/securedrop/tests/functional/test_submit_and_retrieve_message.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/circleci/project/securedrop/tests/functional/journalist_navigation_steps.py:798: in _journalist_downloads_message
    self._journalist_selects_the_first_source()
/home/circleci/project/securedrop/tests/functional/journalist_navigation_steps.py:790: in _journalist_selects_the_first_source
    self.driver.find_element_by_css_selector("#un-starred-source-link-1").click()
[...]
E       selenium.common.exceptions.ElementNotInteractableException: Message: Element <a id="un-starred-source-link-1" class="code-name unread" href="/col/E7DV7R4U3FALMXE3N2THUA7DQQUPAUUKFOENZPO7FGEGRELBC3XQM22KMMR53TOJFAN3XILJ6FQTZGRDG6DDNRXXF66JXBXAYMJXABA="> could not be scrolled into view

/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py:242: ElementNotInteractableException
```

This behavior is not reproducible in manual testing (in Chrome, Firefox, or Tor Browser) and in 20546031a64d647926bff7f011c4a92ab7240600 persists even if `JournalistNavigationStepsMixin._journalist_selects_the_first_source()` uses `FunctionalTest.safe_click_by_id()` (as it probably should).  I've tinkered quite a bit with the test logic, to no avail, and will now poke at the markup to see if I can understand what Selenium is complaining about.  Per eb9f35da08cabd0001bac29120c12bb21db7cb48, it doesn't *appear* to be the new overlaid `dialog` elements.

SaptakS commented on securedrop/journalist_templates/_source_row.html:
```diff
@@ -1,44 +1,58 @@
-  <time class="date" title="{{ source.last_updated|rel_datetime_format }}" datetime="{{ source.last_updated|html_datetime_format }}">{{ source.last_updated|rel_datetime_format(relative=True) }}</time>
-  <div class="designation">
+<tr class="source {% if source.num_unread != 0 %}unread{% else %}read{% endif %}" data-source-designation="{{ source.journalist_designation|lower }}">
+  <td aria-label="{{ source.last_updated|rel_datetime_format(relative=True) }}"><time class="date" title="{{ source.last_updated|rel_datetime_format }}" datetime="{{ source.last_updated|html_datetime_format }}">{{ source.last_updated|rel_datetime_format(relative=True) }}</time></td>
```
I don't see the need for an aria-label here, since the inner html already contains the time element with the same text content.

cfm replied:
I agree in theory!  In practice, at least my screen-reader (VoiceOver on macOS) reads the `datetime` attribute instead of the inner HTML, unless overridden with this `aria-label`.

SaptakS commented on securedrop/journalist_templates/account_edit_hotp_secret.html:
```diff
@@ -1,11 +1,10 @@
-  </p>
-  <button type="submit">{{ gettext('CONTINUE') }}</button>
+  <label for="otp_secret">{{ gettext('Change Secret') }}</label>
+  <input name="otp_secret" type="text" placeholder="{{ gettext('HOTP Secret') }}">
```
There needs to be an `id="otp_secret"` other the for attribute in label doesn't really make sense. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#attr-for

cfm replied:
Absolutely; this was sloppy on my part.  Fixed in f54157a73.

SaptakS commented on securedrop/journalist_templates/account_edit_hotp_secret.html:
```diff
@@ -1,11 +1,10 @@
-  <button type="submit">{{ gettext('CONTINUE') }}</button>
+  <label for="otp_secret">{{ gettext('Change Secret') }}</label>
+  <input name="otp_secret" type="text" placeholder="{{ gettext('HOTP Secret') }}">
+  <button type="submit" aria-label="{{ gettext('Continue') }}">{{ gettext('CONTINUE') }}</button>
```
Adding an `aria-label` here may make sense in cake a more verbose text is added. The idea is if links and buttons have similar text throughout the website, e.g. Continue/cancel might be used in different places, it's better to have an `aria-label` that says continue to what. Something like `Continue to change secret` or something better. Just to convert from all caps doesn't make a lot of sense to me.

cfm replied:
Thanks.  I've expanded a number of `aria-label`s in 39954528f.  In other places, the change of capitalization remains to prevent screen-readers from mistaken all-caps words for initialisms.

SaptakS commented on securedrop/journalist_templates/account_new_two_factor.html:
```diff
@@ -11,8 +11,8 @@ <h1>{{ gettext('Enable FreeOTP') }}</h1>
-<p>{{ gettext("Can't scan the barcode? You can manually pair FreeOTP with your SecureDrop account by entering the following two-factor secret into the app:") }}
-<p class="center"><span id="shared-secret">{{ user.formatted_otp_secret }}</span></p>
+<p>{{ gettext("Can't scan the barcode? You can manually pair FreeOTP with your SecureDrop account by entering the following two-factor secret into the app:") }}</p>
+<mark id="shared-secret" class="center">{{ user.formatted_otp_secret }}</mark>
```
👍🏽

SaptakS commented on securedrop/journalist_templates/account_new_two_factor.html:
```diff
@@ -22,6 +22,6 @@ <h1>{{ gettext('Enable YubiKey (OATH-HOTP)') }}</h1>
   <label for="token">{{ gettext('Verification code') }}</label>
   <input name="token" id="token" type="text" placeholder="123456">
-  <button type="submit">{{ gettext('SUBMIT') }}</button>
+  <button type="submit" aria-label="{{ gettext('Submit') }}">{{ gettext('SUBMIT') }}</button>
```
Same as mentioned above since `Submit` is also pretty common text that appears repeatedly.

SaptakS commented on securedrop/journalist_templates/admin.html:
```diff
@@ -2,57 +2,67 @@
     {{ gettext('ADD USER') }}
   </button>
 </form>
+</section>
```
I feel this entire section can just be an `<a>` tag. Clicking the `Add User` takes the user to a separate page, so I think that qualifies as a link. Don't see the point of a form with button, unless, I am missing something.

cfm replied:
Great catch; thank you.  I'm not sure why 54834e22309e3c8dfaead2a07a650a9af87e7ca1 added this `button` rather than an `a.btn`, which is what 2be57f94ea50683c5a7a3d63c67f25a3097219bf subsequently did for the **Instance Config** button.  Fixed in f4d25065d6d3a74bbbbae7e264f91173389de3f9.

SaptakS commented on securedrop/journalist_templates/admin.html:
```diff
@@ -78,14 +88,16 @@ <h1>{{ gettext('Admin Interface') }}</h1>
 </a>
-
-<hr class="no-line">
+</section>
```
Why convert it into a section instead of just an `<a>` tag? You can add the text inside h2 as the `aria-label` for the anchor tag since it's a bit verbose.

cfm replied:
I appreciate this simplification—done in f4d25065d6d3a74bbbbae7e264f91173389de3f9.

SaptakS commented on securedrop/journalist_templates/admin_add_user.html:
```diff
@@ -1,80 +1,79 @@
+          <span class="form-validation-error" role="alert">{{ error }}</span>
         {% endfor %}
-      </p>
+      </div>
```
Don't think this `<div>` is needed. There are two consecutive `<div>`s instead of just one.

cfm replied:
Thanks!  I'd misread this as necessary for the flexbox layout.  Fixed in cbc3b8ace.

SaptakS commented on securedrop/journalist_templates/admin_add_user.html:
```diff
@@ -1,80 +1,79 @@
+          <span class="form-validation-error" role="alert">{{ error }}</span>
         {% endfor %}
-      </p>
+      </div>
```
Same as above.

SaptakS commented on securedrop/journalist_templates/admin_edit_hotp_secret.html:
```diff
@@ -1,12 +1,11 @@
-  </p>
-  <button type="submit">{{ gettext('CONTINUE') }}</button>
+  <label for="otp_secret">{{ gettext('Change Secret') }}</label>
+  <input name="otp_secret" type="text" placeholder="{{ gettext('HOTP Secret') }}">
```
Needs `id` attribute for `for` to work.

cfm replied:
Also fixed in f54157a73.

SaptakS reviewed (commented):
This looks awesome! Left few comments that I think should be addressed. Some of the comments are applicable in multiple places and templates (e.g., missing id attribute and aria-label on buttons comments ). The comments are related to only HTML semantics, since I think the styling is still under progress and there are some "Fix me" sections.

cfm committed:
adds missing ID attributes for LABEL[for]

thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766762160
thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766786575

cfm committed:
expands ARIA-LABELs to add context to recurring UI elements

thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766764884
thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766766877

cfm committed:
refactors unnecessary FORM/SECTIONs to standalone A.btn elements

thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766769512
thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766770861

cfm committed:
removes unnecessary DIVs

fixup: f3593e0bcd3a6aa48911b8888741cb721faa70f0
thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766781073
thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r766781232

cfm commented:
Thanks, @SaptakS!  I've been able to resolve all but one of your suggestions right away.  I appreciate your careful eye on these changes!  On now to CSS polishing.

cfm committed:
avoids unnecessarily introducing a new source string

"HOTP Secret" already exists as a translated source string, so there's
no reason to introduce "OTP Secret" in the same context.

cfm committed:
third pass through "/login"

This reverts commit 6d4602da288b398dea3680dfb9faba67cf11e85a.

cfm committed:
third pass through "/": fixes "index.html" and "_source_row.html" styling

NB. The refactoring of the UL#cols listing into the TABLE#collections
table is not pixel-perfect, because the latter uses strict columnar
spacing.  This is up for debate.

cfm committed:
third pass through "/": fixes "_sources_confirmation*_modal.html" styling

cfm committed:
third pass through "/col/<filesystem_id>": fixes "col.html" styling

NB. As in "index.html" and "_source_row.html", the conversion of
UL.submissions into TABLE.submissions is not pixel-perfect.  In this
case, I've not even attempted to make it so: this layout is to my eye
more idiomatic visually as well as semantically.  This too is up for
debate.

cfm committed:
third pass through "/col/<filesystem_id>": fixes "_confirmation_modal.html" styling

cfm committed:
third pass through "/account/account": fixes "edit_account.html" styling

Of note, we rescope some rules in "_button.sass" to accommodate the
special case of BUTTON[data-tooltip].  These enhancements layer
less elegantly than I'd like, in both markup and stylesheet, which
will be the subject of a forthcoming ticket.

cfm committed:
third pass through "/admin": fixes "admin.html" styling

cfm committed:
third pass through "/admin/config": fixes "config.html" styling

cfm committed:
third pass through "/admin/add": fixes "admin_add_user.html" styling

Along the way, makes spacing more consistent with "/account/account" and
"/admin/edit" too.  However, the visible H1 is a deliberate departure
from the current styling.

cfm committed:
third pass through "/{account,admin}/2fa": fixes "{account,admin}_new*_two_factor.html" styling

cfm committed:
third pass through "/{account,admin}/reset-2fa-secret": fixes "{account,admin}_edit_hotp_secret.html" styling

cfm committed:
fixes "[preferences_saved_]flashed.html" styling

cfm commented:
This pull request's _substantive_ changes are complete with the conclusion of CSS polishing in 07f9b57f.  Remaining work per #5987:

1. get tests passing;
2. do another round of functional and visual testing (Tor Browser) and validation;
3. reformat `*.html` à la 8f2b5031f9ae1f55586a7eeab52cf4503538ee02.

cfm committed:
third pass through "/admin": fixes tests

cfm committed:
third pass through "/col/<filesystem_id>": fixes tests

cfm committed:
fixes regressed NAV placement

cfm committed:
corrects placement of elements that accompany FORM INPUTs

cfm committed:
prevents wrapping of THs in collections, submissions, and users TABLEs

zenmonkeykstop reviewed (changes requested):
- [x] HTML validation (in Tor Browser, via the "validate local html" option of the Web Developer plugin looks good.
- [x] Page is readable via Tails' screen reader
- [x] No visual discrepancies that stood out vs the demo site
- [ ] *FAIL* The Source Interface seems mostly unchanged except for one point:
  - the **Get started** and **Log in** buttons are truncated, and text is not centered in them. (this is relatively major)
  - (for other reviewers, the lock icon on the generate page has been removed in develop, and the "forgot your codename" link in flash messages is on its own line now. These are not issues in this PR, just the demo site lagging behind the develop branch)

I'll poke about a bit to see what's affecting the button padding on the SI, but this looks OK to merge  once that's fixed.

cfm commented:
<https://github.com/freedomofpress/securedrop/pull/6165#pullrequestreview-851972869>:

> [ ] _FAIL_ The Source Interface seems mostly unchanged except for one point:
> * the **Get started** and **Log in** buttons are truncated, and text is not centered in them. (this is relatively major)

Ack; thanks, @zenmonkeykstop.  I'd fixed this at some point but must have lost track of it in a stash or rebase.  Should be a quick fix to reconstruct (or just dig out of my local reflog).

cfm commented:
@zenmonkeykstop, this should be fixed in cbfc08f.  Once you've reviewed and approved, I'll start in (either today or Monday) on the "After Approval and Before Merge" checklist above.

zenmonkeykstop commented:
Confirmed that the SI fix looks good. If you wanna knock off the first two items on that list and leave the squash merge to the approver, I can take a final pass once they're in place.

legoktm commented on securedrop/journalist_templates/base.html:
```diff
@@ -19,40 +17,40 @@
+        {{ gettext("Powered by <b>SecureDrop {version}</b>.").format(version=version)|safe }}
       </footer>
       {% endblock %}
+      {% include 'js-strings.html' %}
```
nit: I think this shouldn't be in the div and just directly right before `</body>`

cfm replied:
Good call, @legoktm; thanks.  Fixup in 340938556.

legoktm commented on securedrop/journalist_templates/edit_account.html:
```diff
@@ -61,8 +67,14 @@ <h2>{{ gettext('Reset Password') }}</h2>
-  <p><input type="password" name="current_password" placeholder="{{ gettext('Current Password') }}"></p>
-  <p><input name="token" id="token" type="text" placeholder="{{ gettext('Two-factor Code') }}"></p>
+  <div>
+    <label for="current_password" class="visually-hidden">{{ gettext('Current Password') }}</label>
```
Not sure if it's in scope, but these labels seem like they should be shown for all users instead of relying solely on the placeholder text.

cfm replied:
I have mixed feelings about the label/placeholder usage across SecureDrop forms too.  In the Journalist Interface:

* labels are visible everywhere fields require explication (e.g., `admin_add_user.html`);
* labels are visible and augmented with placeholders where fields have logical relations (e.g., label "Is using a Yubikey" + placeholder "HOTP Secret"); and
* labels are visually hidden in favor of placeholders, but still present in the accessibility tree, where fields have logical relations familiar enough to be left implicit (e.g., `login.html`'s three fields and the two fields you've cited in `edit_account.html`).

In this case in `edit_account.html`, it seems to me that the two fields here:

1. visually imitate the three fields of the login form;
2. have explanatory text nearby in— <https://github.com/freedomofpress/securedrop/blob/cbfc08f73c56afa970fbdadf8ecbffe4d6d2e7e3/securedrop/journalist_templates/edit_account.html#L65>—and
3. still benefit from their placeholder text.

So, for kicks, here is c9eb580a74e9b066d71724a5845c37d1aa6342d7, where these fields are grouped in a `fieldset` with an explanatory `legend`, otherwise keeping the `login.html`-like field layout.  I'll be curious to know whether you and @zenmonkeykstop like this or think it's more trouble than it's worth.  :-)

legoktm replied:
I like the fieldset, it's a minimal change but IMO improves the usability of it.

cfm committed:
groups related fields in a FIELDSET with an explanatory LEGEND (per review feedback)

thread: https://github.com/freedomofpress/securedrop/pull/6165#discussion_r785170736

cfm committed:
formats templates for more-consistent indentation

cfm commented:
Re: the pre-merge step:

> Update as many translations as possible per <https://github.com/freedomofpress/securedrop/issues/5972#issuecomment-947317684>; test in Weblate sandbox

Unfortunately, after running `make translate` on this branch, the only developer-friendly translation drift I see is (e.g., for `ar`):

```patch
-msgid "Powered by <em>SecureDrop {version}</em>."
+#, fuzzy
+#| msgid "Powered by <em>SecureDrop {version}</em>."
+msgid "Powered by <b>SecureDrop {version}</b>."
 msgstr "من <em>SecureDrop {version} </em>."
```

I'll be happy to do an `s/em/b` pass in Weblate as usual at the start of the translation period.  In the current translation workflow, however, I don't think it's worth trying to be clever about updating the `.po`s directly here for a single HTML fix.

legoktm reviewed (dismissed):
The changes made in response to my feedback LGTM. I did a final skim as well, but not any full testing (I assume Kev already did or will do that)

cfm commented:
Thanks, @legoktm!  I've rebased out the `WIP` commit message and a few `fixup`s.  With that, this is ready for @zenmonkeykstop's final review and merge.

zenmonkeykstop commented:
Thanks all, taking a final spin through now.

zenmonkeykstop reviewed (approved):
Took a quick spin through validation and screen-reader steps again, no worries on that score.
 Source Interface glitch is resolved, looks consistent overall.

LGTM - thanks @cfm for the work and @SaptakS and @legotm for review!
